processor/src/main/java/org/mapstruct/ap/internal/model/presence/AllPresenceChecksPresenceCheck.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.internal.model.presence;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+import org.mapstruct.ap.internal.model.common.ModelElement;
+import org.mapstruct.ap.internal.model.common.PresenceCheck;
+import org.mapstruct.ap.internal.model.common.Type;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class AllPresenceChecksPresenceCheck extends ModelElement implements PresenceCheck {
```

Comment:
From this I assume you want to concatenate multiple presence checks.. But as mentioned earlier, there could be a dependency: a presence checker not being executed while a previous one already fired.

I'm not sure if we need that.. E.g.: can you skip the null presence check?  Or couldn't that one be the default to override (instead of add a check after that). ?

Reply:
How much is this all steered by the strategy enums on method-class-mapper level?

Reply:
Should we deprecate the strategy enums.. now we have more fine grained control.. ? E.g. should this be 2.0 functionality?

Reply:
I don't quite follow with what you mean by "a presence checked not being executed while a previous one already fired". This `AllPresenceChecksPresenceCheck` is meant mostly to cover the use case in #1456.

>How much is this all steered by the strategy enums on method-class-mapper level?

This is entirely steared by the strategy enums. All the behaviour stays the same, basically `NullValueCheckStrategy` can be though as `PresenceValueCheckStrategy`. 

>Should we deprecate the strategy enums.. now we have more fine grained control.. ? E.g. should this be 2.0 functionality?

I don't think that we should deprecate them. I think that it should behave the same way as it did before. With this we are giving more options for the users to write presence checks without changing their beans.

Reply:
> I don't quite follow with what you mean by "a presence checked not being executed while a previous one already fired". This `AllPresenceChecksPresenceCheck` is meant mostly to cover the use case in #1456.

I suspected that the following code is generated:

```java
if ( source != null && pc1( source ) && pc2( source) {
        target.setName( source.getName() );
 }

```

The value `null` will never be offered to the `pc1`. So `pc1` cannot make choices anymore on `null`.. but reading your answer that does not happen right? The same is valid for `pc1` and `pc2` by the way, but at least that is something that the user has under control.

Reply:
> This is entirely steared by the strategy enums. All the behaviour stays the same, basically `NullValueCheckStrategy` can be though as `PresenceValueCheckStrategy`.

What about `NullValuePropertyMappingStrategy`..? 

The point is that this interface is getting too complex. You need to define a strategy and then you need to do something in your `@Mapping` to customise this.

I would plea for a solution that we 
1. make behaviour consistent:.  `NullValuePropertyMappingStrategy' and 'NullValueCheckStrategy' 
2. drop the 2 `NullValuePropertyMappingStrategy' and 'NullValueCheckStrategy' 
3. extend `@Condition` to be used on `@MapperConfig` and `@Mapper`
4. have some default `@Conditions` for reuse. Like the `null check` and the `has` check.
5. have a script to apply to source code migrate the current two strategies. I'm thinking about the way that assertj was done from its predecessor. 
6. Rethink the remaining strategy `NullValueMappingStrategy` and at least call it something different: `SourceBeanNullCheck` or apply a special case of the `@Condition` to this scenario as well.

Don't get me wrong.. I really like the approach.  But I think the API is becoming too complex and too dependent. 

So perhaps we should call the next release 2.0 alpha or something like that.

Reply:
I see what you are saying. In this situation we are not looking for custom presence checkers here, this will work exactly like it did before. The code you think that is generated with `pc1` and `pc2` it won't be generated.

I am not following your plea, perhaps we need to do a sync meeting about it 😉